client/js/widgets/button.js

```diff
@@ -69,10 +69,20 @@ export class Button extends Widget {
       if(a.applyVariables) {
         if(Array.isArray(a.applyVariables)) {
           for(const v of a.applyVariables) {
-            if(v.parameter && v.variable)
-              a[v.parameter] = variables[v.variable];
-            else
-              problems.push('Entry in parameter applyVariables does not contain "parameter" and "variable".');
+            if(v.parameter && v.variable) {
+              a[v.parameter] = (v.index === undefined) ? variables[v.variable] : variables[v.variable][v.index];
+            } else if (v.parameter && v.template) {
+              for (let key in variables) {
+                a[v.parameter] = v.template.replace(/({([^}]+)})/g, function(i) {
+                  let key = i.replace(/{/, '').replace(/}/, '');
```

Comment:
I'm pretty sure you don't need this because you could read the `key` from the RegEx capture groups. The callback gets those as parameters: `function(i, group1, group2, ...)`.

Do these expressions even work though? Don't you have to escape `{` and `}`? Hmm, just tested it and looks like it works. Weird.

Reply:
I think it works because there is no pattern to match immediately before the `{` character - in absence of a valid meta meaning for the character, it is being taken literally.

I'd have written this as `replace(/\{([^}]+)\}/g` escaping the braces to be clear they are being taken literally and only capturing the part between the braces to eliminate the need to remove them on the next line.

Reply:
thanks, that's better